### ISSUE TITLE
combine not honoring comments location

### ISSUE THREAD

**Issue by SID800** [Kintyre/ksconf#66]
# The problem

ksconf combine moves comment lines around in the target files. This makes it challenging to understand the comments intentions.  It also reorders (sort?) the stanzas, which can sound like a good idea... until that alters the logical flow of the code within a .conf file.  Yes, I realize it does not change how Splunk parses/combines the configs, but often developers place things in a certain order on purpose. 

## Environment

* Ksconf version:  ksconf 0.7.7rc3 (Build 449)
* OS & version used: windows
* Python version: 2.7.16 (c:\python27\python.exe) 
* Installed via:  pip

## Details

Here is an Splunk_TA_Windows original /default/wmi.conf
![image](https://user-images.githubusercontent.com/12401288/66434292-18832880-e9d7-11e9-874b-a66d2b44b00f.png)

and wmi.conf after ksconf combine (I have no local changes so this should just be a file copy).
![image](https://user-images.githubusercontent.com/12401288/66434343-394b7e00-e9d7-11e9-9b60-aca3c6bd0194.png)

Note how the comments related to [WMI:LocalApplication] were moved under [settings] which would lead an admin astray.

**Comment by lowell80** [comment]
Yes, this is a known limitation.  The solution requires re-writing the parser so that comments are no longer just second-class citizens.   I mean, they will always be "lesser" because they don't have any functional impact, but they should be preserved, where possible.

## Where to comments belong?

Sometimes it's impossible to always know which what entity a comment should be associated to. Take for example:

```
# Global comment

# Thoughts about my stanza
[cool_stanza]
attribute1 = x
# comment about attr2
attr = some complex thing
....
```

How can the parser know if a comment goes with the stanza before or after it?   How do you know if a comment should be pinned to an attribute?  It will never be fully correct.

I eventually want the option to (where possible) preserve stanza ordering and/or key ordering on a file-by-file basis.   Specifically props.conf and transforms.conf are often managed by hand, where comments greatly enhance reliability.  But for now pretty much everytime ksconf writes to a`.conf` file, that file is sorted (and normalized); (This is not true for `ksconf unarchive`).  Typically this is a good thing, and some places this will be unavoidable, but I want it to be more flexible in the future.


## Current behavior

Right now the behavior is:

1.  Stanzas are sorted by their name.
2. Comments always go with the stanza above it.   (I've gotten decent leverage by just keeping all my comments at the very top of the stanzas; and if I'm commenting on an attribute I'm careful to NAME the attribute at the front of the comment.)  Global comments are normally handled okay typically, because they are considered to be part of the default/global/top-level stanza.
3.  Attribute names are sorted.
4.  Comment order is preserved (per stanza); but comments mixed with attributes are not kept.
5.  Sorting can be blocked on individual files by using the `KSCONF-NO-SORT` magic "marker".  But this only works for the `sort` command, anytime ksconf has to write to a conf file, the result will be a sorted file.

## Preserving order with complex commands

Certain operations, like merge, combine, or promote are very likely never going to support preservation of comments.  And the reason is simply that, there's basically no way to get it "right".  The problem above of figuring out where a comment belongs, is just multiplied even further when you add more files into.   When merging comments, should identical comments be suppressed and just outputted once?  What if an identical lines are in a different order between two files?  What if the same comment is present in 3 out of 4 files, and the 4th file is the highest precedence.  Should that comment be removed?

The primary purpose of these higher-order tools is to provide output that's usable by Splunk; not humans.

As much as it annoys me to have my comments screwed up; I've learned to live with it (at least for now).  I haven't been able to justify prioritizing the effort required to properly fix this over all the other things that make a functional impact.

Any conf file that Splunk edits is subject to the same potential chaotic update.

## TL;DR

1. Yes it's a limitation
2. Your comments will be kept with the correct stanza if you move them down into the body of a stanza rather than before it.  (It felt weird to me at first, but I got used to it.  Best option for now.)
3.  It may be fixed when the core parser is reworked, however:
4.  Don't expect it to be available for *all* commands.

**Comment by lowell80** [comment]
Reviewing tickets and just realized that I didn't point out the other obvious solution.   Just place the string `KSCONF-NO-SORT` in a comment at the top of the file that that will prevent some commands from touching it, like `ksconf sort`.  I used this for files that I really want to maintain myself, which is often `props.conf` for me.   I like to add example log messages along side the `EXTRACT-*` entries so that I can see the different variations as I have to update the regexes over time.  Also, `indexes.conf` and `searchbnf.conf`....
